pkg/network/ebpf/c/protocols/postgres/decoding.h

```diff
@@ -226,33 +246,47 @@ static __always_inline bool handle_response(pktbuf_t pkt, conn_tuple_t conn_tupl
     }
 
 #pragma unroll(POSTGRES_MAX_MESSAGES_PER_TAIL_CALL)
-    for (__u32 iteration = 0; iteration < POSTGRES_MAX_MESSAGES_PER_TAIL_CALL; ++iteration) {
-        if (!read_message_header(pkt, &header)) {
+    for (__u8 messages_count = 0; messages_count < POSTGRES_MAX_MESSAGES_PER_TAIL_CALL; ++messages_count) {
+        read_result = read_message_header(pkt, &header);
+        if (read_result != READ_OK) {
             break;
         }
         if (header.message_tag == POSTGRES_COMMAND_COMPLETE_MAGIC_BYTE) {
-            handle_command_complete(&conn_tuple, transaction);
             found_command_complete = true;
             break;
         }
         // We didn't find a command complete message, so we advance the data offset to the end of the message.
         // reminder, the message length includes the size of the payload, 4 bytes of the message length itself, but not
         // the message tag. So we need to add 1 to the message length to jump over the entire message.
         pktbuf_advance(pkt, header.message_len + 1);
+        iteration_value->msg_count++;
     }
 
+    if (found_command_complete || (read_result == READ_FRAGMENTED)) {
+        // if completion was found or the packet was fragmented, then report the transaction and clear it from the map.
+        handle_command_complete(&conn_tuple, transaction);
+    }
     if (found_command_complete) {
+        // save messages counter in bucket.
+        __u8 bucket_idx = iteration_value->msg_count / PG_KERNEL_MSG_COUNT_BUCKET_SIZE;
+        bucket_idx = (bucket_idx >= PG_KERNEL_MSG_COUNT_NUM_BUCKETS) ? (PG_KERNEL_MSG_COUNT_NUM_BUCKETS - 1) : bucket_idx;
+        __sync_fetch_and_add(&pg_msg_counts->pg_messages_count_buckets[bucket_idx], 1);
+
         return 0;
     }
-    // We didn't find a command complete message, so we need to continue processing the packet.
-    // We save the current data offset and increment the iteration counter.
-    iteration_value->iteration += 1;
-    iteration_value->data_off = pktbuf_data_offset(pkt);
+    if ((read_result == READ_FRAGMENTED) || (iteration_value->msg_count == (PG_KERNEL_MAX_MESSAGES - 1))) {
+        // the packet was fragmented or max messages - add counter to the last bucket and stop iterating.
+        __sync_fetch_and_add(&pg_msg_counts->pg_messages_count_buckets[PG_KERNEL_MSG_COUNT_NUM_BUCKETS], 1);
 
-    // If the maximum number of tail calls has been reached, we can skip invoking the next tail call.
-    if (iteration_value->iteration >= POSTGRES_MAX_TAIL_CALLS_FOR_MAX_MESSAGES) {
         return 0;
     }
```

Comment:
why did you remove this check?
What code piece promise we are executing up to `POSTGRES_MAX_TAIL_CALLS_FOR_MAX_MESSAGES` times?

Reply:
[Line 277 ](https://github.com/DataDog/datadog-agent/blob/aed60c21f5edccf132acc4b414a512231ce358d7/pkg/network/ebpf/c/protocols/postgres/decoding.h#L277) stops further iteration is reached allowed max messages

Reply:
I've changed condition to stop iterator